### PR TITLE
Fix covariant returns issue with partially canonical instantiation

### LINKED ISSUE #47007: Still getting System.TypeLoadException with generics and C# covariant return types
This is a continuation of https://github.com/dotnet/runtime/issues/45344 (which has had discussion closed). The issue is still persisting for me on 5.0.102.

Repro:

```
_ = new Derived<string, int>();

class Base<T1>
{
    public virtual Base<T1> Method() => null;
}
class Derived<T1, T2> : Base<T1>
{
    public override Derived<T1, T2> Method() => null;
}
```

Note that this does work if doing any other combination of `string`/`int` as type arguments into `Derived<,>`.

EDIT: Apologies, I must have been looking at another issue when I thought the original was closed off - let me know if I should copy this into it instead and let this one get shut.

### PR DESCRIPTION
**Pull request by janvorli**
The ClassLoader::ValidateMethodsWithCovariantReturnTypes was properly
handling canonical instantiations only when all the type parameters were
canonical. But when only some of them were cannonical, it was not
getting the non-canonical version of the instantiation and the
validation was failing.

This change fixes it by using ClassLoader::IsCanonicalGenericInstantiation
instead of the IsTypicalSharedInstantiation.

I've also added a regression test for the problem and added couple of
missing licence headers to regression tests I have recently created.

Close #47007

### CODE REVIEW

File: src/coreclr/vm/class.cpp [hunk 1]
```diff
@@ -1170,7 +1170,7 @@ void ClassLoader::ValidateMethodsWithCovariantReturnTypes(MethodTable* pMT)
                 continue;
 
             Instantiation parentClassInst = pParentMD->GetClassInstantiation();
-            if (ClassLoader::IsTypicalSharedInstantiation(parentClassInst))
+            if (ClassLoader::IsCanonicalGenericInstantiation(parentClassInst))
```
Comment (davidwrighton) [inline src/coreclr/vm/class.cpp:1173]: Is there a reason why this isn't if (true)?

Comment (janvorli) [reply src/coreclr/vm/class.cpp:1173]: I am not sure I understand the question.

Comment (janvorli) [reply src/coreclr/vm/class.cpp:1173]: Ah, you mean that for non-canonical generic instantiation in pParentMD->GetClassInstantiation(), the pParentMD->GetClassInstantiation()==pParentMT->GetInstantiation(), so we never need to check.

Comment (janvorli) [reply src/coreclr/vm/class.cpp]: @davidwrighton I've removed the `if` here and also below for the current class instantiation. All the tests are passing with that change.

**Review by davidwrighton — APPROVED** [review]

**Pull request merged**